core/src/nucleus/ribosome/api/query.rs

```diff
@@ -0,0 +1,38 @@
+use holochain_core_types::{
+    entry_type::EntryType,
+};
+use holochain_wasm_utils::api_serialization::{QueryArgs, QueryResult};
+use nucleus::ribosome::Runtime;
+use serde_json;
+use std::str::FromStr;
+use wasmi::{RuntimeArgs, RuntimeValue, Trap};
+
+/// ZomeApiFunction::query function code
+/// args: [0] encoded MemoryAllocation as u32
+/// Expected complex argument: ?
+/// Returns an HcApiReturnCode as I32
+pub fn invoke_query(
+    runtime: &mut Runtime,
+    args: &RuntimeArgs,
+) -> Result<Option<RuntimeValue>, Trap> {
+    // deserialize args
+    let args_str = runtime.load_utf8_from_args(&args);
+    let input: QueryArgs = match serde_json::from_str(&args_str) {
+        Ok(input) => input,
+        Err(_) => return ribosome_error_code!(ArgumentDeserializationFailed),
+    };
+    // Get entry_type
+    let maybe_entry_type = EntryType::from_str(&input.entry_type_name);
+    if maybe_entry_type.is_err() {
+        return ribosome_error_code!(UnknownEntryType);
+    }
+    let entry_type = maybe_entry_type.unwrap();
```

Comment:
Better do

```rust
let entry_type = EntryType::from_str(&input.entry_type_name)
    .map_err(|_| ribosome_error_code!(UnknownEntryType))?;
```
It's much cleaner

Reply:
Sadly, that doesn’t work. ˋribosome_error_code!ˋ does not return an Err but an Ok..

Reply:
Okay, then that's wrong. An error is an error. An ok value can't be an error

Reply:
well yes it can be

ok means "the zome ran" not "the result of the zome had no errors"

Reply:
would also be out of scope to change that

Reply:
I agree, but we should discuss this. This is logically and semantically wrong.

Reply:
It did confuse me already.

Reply:
I designed I new error system: https://drive.google.com/drive/u/1/folders/12VqxGg_kar2mW_HESP9kKAQnJOuT-qBy

How's that for a discussion? :)